File: BookManager.cs
// BookManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

public record Book(
    int Id,
    string Title,
    string Author,
    int Year,
    string Genre,
    bool Read,
    int Rating,
    string AddedDate
);

public class BooksData
{
    public List<Book> Books { get; set; } = new();
}

public class BookManager
{
    private List<Book> books = new();
    private int nextId = 1;

    public IReadOnlyList<Book> Books => books.AsReadOnly();

    public Book AddBook(string title, string author, int year, string genre, bool read, int rating)
    {
        if (rating < 1 || rating > 10) throw new ArgumentException("Оценка должна быть от 1 до 10");
        if (year < 0 || year > DateTime.Now.Year)
            throw new ArgumentException($"Год должен быть от 0 до {DateTime.Now.Year}");
        var book = new Book(nextId, title, author, year, genre, read, rating, DateTime.Now.ToString("yyyy-MM-dd"));
        books.Add(book);
        nextId++;
        return book;
    }

    public Book? FindBook(int id) => books.FirstOrDefault(b => b.Id == id);

    public bool EditBook(int id, Dictionary<string, object> updates)
    {
        var old = FindBook(id);
        if (old == null) return false;
        books.Remove(old);
        string title = updates.ContainsKey("title") ? (string)updates["title"] : old.Title;
        string author = updates.ContainsKey("author") ? (string)updates["author"] : old.Author;
        int year = updates.ContainsKey("year") ? (int)updates["year"] : old.Year;
        string genre = updates.ContainsKey("genre") ? (string)updates["genre"] : old.Genre;
        bool read = updates.ContainsKey("read") ? (bool)updates["read"] : old.Read;
        int rating = updates.ContainsKey("rating") ? (int)updates["rating"] : old.Rating;
        var updated = new Book(old.Id, title, author, year, genre, read, rating, old.AddedDate);
        books.Add(updated);
        return true;
    }

    public bool DeleteBook(int id) => books.RemoveAll(b => b.Id == id) > 0;

    public List<Book> SearchBooks(string query)
    {
        var q = query.ToLower();
        return books.Where(b => b.Title.ToLower().Contains(q) || b.Author.ToLower().Contains(q)).ToList();
    }

    public List<Book> FilterByRead(bool read) => books.Where(b => b.Read == read).ToList();

    public List<Book> FilterByGenre(string genre) =>
        books.Where(b => string.Equals(b.Genre, genre, StringComparison.OrdinalIgnoreCase)).ToList();

    public Dictionary<string, object> GetStats()
    {
        int total = books.Count;
        int readCount = FilterByRead(true).Count;
        int unread = total - readCount;
        double avgRating = FilterByRead(true).Any() ? FilterByRead(true).Average(b => b.Rating) : 0;
        var genres = books.GroupBy(b => b.Genre).ToDictionary(g => g.Key, g => g.Count());
        return new Dictionary<string, object>
        {
            ["total"] = total,
            ["read"] = readCount,
            ["unread"] = unread,
            ["avg_rating"] = avgRating,
            ["genres"] = genres
        };
    }

    public void SaveToFile(string filename)
    {
        var data = new BooksData { Books = books };
        var options = new JsonSerializerOptions { WriteIndented = true };
        string json = JsonSerializer.Serialize(data, options);
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        var data = JsonSerializer.Deserialize<BooksData>(json);
        if (data != null)
        {
            books = data.Books;
            nextId = books.Any() ? books.Max(b => b.Id) + 1 : 1;
        }
    }
}

public static class Program
{
    private static string ReadString(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private static int ReadInt(string prompt)
    {
        while (true)
        {
            Console.Write(prompt);
            if (int.TryParse(Console.ReadLine(), out int result))
                return result;
            Console.WriteLine("Введите число.");
        }
    }

    private static bool ReadBool(string prompt)
    {
        while (true)
        {
            string input = ReadString(prompt);
            if (input == "1") return true;
            if (input == "0") return false;
            Console.WriteLine("Введите 1 или 0.");
        }
    }

    private static void PrintBook(Book book)
    {
        string status = book.Read ? "✅ Прочитана" : "📖 Не прочитана";
        Console.WriteLine($"#{book.Id} - {book.Title} ({book.Year})");
        Console.WriteLine($"   Автор: {book.Author}, Жанр: {book.Genre}");
        Console.WriteLine($"   {status}, Оценка: {book.Rating}/10");
        Console.WriteLine($"   Добавлена: {book.AddedDate}");
    }

    public static void Main()
    {
        var manager = new BookManager();
        try { manager.LoadFromFile("books_data.json"); }
        catch { Console.WriteLine("Не удалось загрузить данные."); }

        while (true)
        {
            Console.WriteLine("\n===== МЕНЕДЖЕР КНИГ (C#) =====");
            Console.WriteLine("1. Добавить книгу");
            Console.WriteLine("2. Показать все книги");
            Console.WriteLine("3. Показать прочитанные книги");
            Console.WriteLine("4. Показать непрочитанные книги");
            Console.WriteLine("5. Найти книги по автору или названию");
            Console.WriteLine("6. Редактировать книгу");
            Console.WriteLine("7. Удалить книгу");
            Console.WriteLine("8. Показать статистику");
            Console.WriteLine("9. Сохранить в файл");
            Console.WriteLine("10. Загрузить из файла");
            Console.WriteLine("0. Выход");
            string choice = ReadString("Выберите действие: ");

            switch (choice)
            {
                case "0": return;
                case "1":
                    string title = ReadString("Название: ");
                    if (string.IsNullOrWhiteSpace(title)) { Console.WriteLine("Название не может быть пустым."); continue; }
                    string author = ReadString("Автор: ");
                    if (string.IsNullOrWhiteSpace(author)) { Console.WriteLine("Автор не может быть пустым."); continue; }
                    int year = ReadInt("Год издания: ");
                    string genre = ReadString("Жанр: ");
                    if (string.IsNullOrWhiteSpace(genre)) genre = "Неизвестно";
                    bool read = ReadBool("Статус (1-прочитана, 0-не прочитана): ");
                    int rating = ReadInt("Оценка (1-10): ");
                    try
                    {
                        var book = manager.AddBook(title, author, year, genre, read, rating);
                        Console.WriteLine($"Книга добавлена с ID {book.Id}");
                    }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "2":
                    if (!manager.Books.Any()) Console.WriteLine("Нет книг.");
                    else foreach (var b in manager.Books) PrintBook(b);
                    break;
                case "3":
                    var readBooks = manager.FilterByRead(true);
                    if (!readBooks.Any()) Console.WriteLine("Нет прочитанных книг.");
                    else foreach (var b in readBooks) PrintBook(b);
                    break;
                case "4":
                    var unreadBooks = manager.FilterByRead(false);
                    if (!unreadBooks.Any()) Console.WriteLine("Нет непрочитанных книг.");
                    else foreach (var b in unreadBooks) PrintBook(b);
                    break;
                case "5":
                    string query = ReadString("Введите часть названия или автора: ");
                    var results = manager.SearchBooks(query);
                    if (!results.Any()) Console.WriteLine("Книги не найдены.");
                    else foreach (var b in results) PrintBook(b);
                    break;
                case "6":
                    int id = ReadInt("Введите ID книги для редактирования: ");
                    var old = manager.FindBook(id);
                    if (old == null) { Console.WriteLine("Книга не найдена."); continue; }
                    Console.WriteLine("Оставьте поле пустым, чтобы не менять.");
                    string newTitle = ReadString($"Название ({old.Title}): ");
                    string newAuthor = ReadString($"Автор ({old.Author}): ");
                    string newYearStr = ReadString($"Год ({old.Year}): ");
                    string newGenre = ReadString($"Жанр ({old.Genre}): ");
                    string newReadStr = ReadString($"Статус (1-прочитана, 0-не прочитана) сейчас: {(old.Read ? "1" : "0")}: ");
                    string newRatingStr = ReadString($"Оценка ({old.Rating}): ");
                    var updates = new Dictionary<string, object>();
                    if (!string.IsNullOrWhiteSpace(newTitle)) updates["title"] = newTitle;
                    if (!string.IsNullOrWhiteSpace(newAuthor)) updates["author"] = newAuthor;
                    if (!string.IsNullOrWhiteSpace(newYearStr))
                    {
                        if (int.TryParse(newYearStr, out int y)) updates["year"] = y;
                        else Console.WriteLine("Год должен быть числом, пропускаем.");
                    }
                    if (!string.IsNullOrWhiteSpace(newGenre)) updates["genre"] = newGenre;
                    if (!string.IsNullOrWhiteSpace(newReadStr)) updates["read"] = newReadStr == "1";
                    if (!string.IsNullOrWhiteSpace(newRatingStr))
                    {
                        if (int.TryParse(newRatingStr, out int r)) updates["rating"] = r;
                        else Console.WriteLine("Оценка должна быть числом, пропускаем.");
                    }
                    if (manager.EditBook(id, updates)) Console.WriteLine("Книга обновлена.");
                    else Console.WriteLine("Ошибка обновления.");
                    break;
                case "7":
                    int delId = ReadInt("Введите ID книги для удаления: ");
                    if (manager.DeleteBook(delId)) Console.WriteLine("Книга удалена.");
                    else Console.WriteLine("Книга не найдена.");
                    break;
                case "8":
                    var stats = manager.GetStats();
                    Console.WriteLine("\n=== СТАТИСТИКА ===");
                    Console.WriteLine($"Всего книг: {stats["total"]}");
                    Console.WriteLine($"Прочитано: {stats["read"]}");
                    Console.WriteLine($"Не прочитано: {stats["unread"]}");
                    Console.WriteLine($"Средняя оценка: {stats["avg_rating"]:F2}");
                    Console.WriteLine("По жанрам:");
                    var genres = (Dictionary<string, int>)stats["genres"];
                    foreach (var kv in genres) Console.WriteLine($"  {kv.Key}: {kv.Value}");
                    break;
                case "9":
                    try { manager.SaveToFile("books_data.json"); Console.WriteLine("Сохранено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                case "10":
                    try { manager.LoadFromFile("books_data.json"); Console.WriteLine("Загружено."); }
                    catch (Exception ex) { Console.WriteLine($"Ошибка: {ex.Message}"); }
                    break;
                default: Console.WriteLine("Неизвестная команда."); break;
            }
        }
    }
}

File: BookManager.java
// BookManager.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

record Book(int id, String title, String author, int year, String genre, boolean read, int rating, String addedDate) implements Serializable {}

class BooksData implements Serializable {
    private static final long serialVersionUID = 1L;
    public List<Book> books;
}

class BookManager implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<Book> books = new ArrayList<>();
    private int nextId = 1;

    public Book addBook(String title, String author, int year, String genre, boolean read, int rating) {
        if (rating < 1 || rating > 10) throw new IllegalArgumentException("Оценка должна быть от 1 до 10");
        if (year < 0 || year > LocalDate.now().getYear())
            throw new IllegalArgumentException("Год должен быть от 0 до " + LocalDate.now().getYear());
        Book book = new Book(nextId, title, author, year, genre, read, rating, LocalDate.now().toString());
        books.add(book);
        nextId++;
        return book;
    }

    public Optional<Book> findBook(int id) {
        return books.stream().filter(b -> b.id() == id).findFirst();
    }

    public boolean editBook(int id, Map<String, Object> updates) {
        Optional<Book> opt = findBook(id);
        if (opt.isEmpty()) return false;
        Book old = opt.get();
        books.remove(old);
        String title = (String) updates.getOrDefault("title", old.title());
        String author = (String) updates.getOrDefault("author", old.author());
        int year = (int) updates.getOrDefault("year", old.year());
        String genre = (String) updates.getOrDefault("genre", old.genre());
        boolean read = (boolean) updates.getOrDefault("read", old.read());
        int rating = (int) updates.getOrDefault("rating", old.rating());
        Book updated = new Book(old.id(), title, author, year, genre, read, rating, old.addedDate());
        books.add(updated);
        return true;
    }

    public boolean deleteBook(int id) {
        return books.removeIf(b -> b.id() == id);
    }

    public List<Book> searchBooks(String query) {
        String q = query.toLowerCase();
        return books.stream()
                .filter(b -> b.title().toLowerCase().contains(q) || b.author().toLowerCase().contains(q))
                .collect(Collectors.toList());
    }

    public List<Book> filterByRead(boolean read) {
        return books.stream().filter(b -> b.read() == read).collect(Collectors.toList());
    }

    public List<Book> filterByGenre(String genre) {
        return books.stream().filter(b -> b.genre().equalsIgnoreCase(genre)).collect(Collectors.toList());
    }

    public Map<String, Object> getStats() {
        int total = books.size();
        int readCount = filterByRead(true).size();
        int unread = total - readCount;
        double avgRating = filterByRead(true).stream().mapToInt(Book::rating).average().orElse(0);
        Map<String, Integer> genres = new HashMap<>();
        books.forEach(b -> genres.put(b.genre(), genres.getOrDefault(b.genre(), 0) + 1));
        Map<String, Object> stats = new HashMap<>();
        stats.put("total", total);
        stats.put("read", readCount);
        stats.put("unread", unread);
        stats.put("avg_rating", avgRating);
        stats.put("genres", genres);
        return stats;
    }

    public void saveToFile(String filename) throws IOException {
        BooksData data = new BooksData();
        data.books = new ArrayList<>(books);
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(data);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            BooksData data = (BooksData) ois.readObject();
            books = new ArrayList<>(data.books);
            for (Book b : books) {
                if (b.id() >= nextId) nextId = b.id() + 1;
            }
        }
    }

    public List<Book> getBooks() { return Collections.unmodifiableList(books); }
}

public class BookManagerApp {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static boolean readBool(String prompt) {
        while (true) {
            String input = readString(prompt);
            if (input.equals("1")) return true;
            if (input.equals("0")) return false;
            System.out.println("Введите 1 или 0.");
        }
    }

    private static void printBook(Book book) {
        String status = book.read() ? "✅ Прочитана" : "📖 Не прочитана";
        System.out.printf("#%d - %s (%d)%n", book.id(), book.title(), book.year());
        System.out.printf("   Автор: %s, Жанр: %s%n", book.author(), book.genre());
        System.out.printf("   %s, Оценка: %d/10%n", status, book.rating());
        System.out.printf("   Добавлена: %s%n", book.addedDate());
    }

    public static void main(String[] args) {
        BookManager manager = new BookManager();
        try {
            manager.loadFromFile("books_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные.");
        }

        while (true) {
            System.out.println("\n===== МЕНЕДЖЕР КНИГ (Java) =====");
            System.out.println("1. Добавить книгу");
            System.out.println("2. Показать все книги");
            System.out.println("3. Показать прочитанные книги");
            System.out.println("4. Показать непрочитанные книги");
            System.out.println("5. Найти книги по автору или названию");
            System.out.println("6. Редактировать книгу");
            System.out.println("7. Удалить книгу");
            System.out.println("8. Показать статистику");
            System.out.println("9. Сохранить в файл");
            System.out.println("10. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    String title = readString("Название: ");
                    if (title.isBlank()) { System.out.println("Название не может быть пустым."); continue; }
                    String author = readString("Автор: ");
                    if (author.isBlank()) { System.out.println("Автор не может быть пустым."); continue; }
                    int year = readInt("Год издания: ");
                    String genre = readString("Жанр: ");
                    if (genre.isBlank()) genre = "Неизвестно";
                    boolean read = readBool("Статус (1-прочитана, 0-не прочитана): ");
                    int rating = readInt("Оценка (1-10): ");
                    try {
                        Book book = manager.addBook(title, author, year, genre, read, rating);
                        System.out.println("Книга добавлена с ID " + book.id());
                    } catch (Exception e) {
                        System.out.println("Ошибка: " + e.getMessage());
                    }
                }
                case "2" -> {
                    if (manager.getBooks().isEmpty()) System.out.println("Нет книг.");
                    else manager.getBooks().forEach(BookManagerApp::printBook);
                }
                case "3" -> {
                    var books = manager.filterByRead(true);
                    if (books.isEmpty()) System.out.println("Нет прочитанных книг.");
                    else books.forEach(BookManagerApp::printBook);
                }
                case "4" -> {
                    var books = manager.filterByRead(false);
                    if (books.isEmpty()) System.out.println("Нет непрочитанных книг.");
                    else books.forEach(BookManagerApp::printBook);
                }
                case "5" -> {
                    String query = readString("Введите часть названия или автора: ");
                    var results = manager.searchBooks(query);
                    if (results.isEmpty()) System.out.println("Книги не найдены.");
                    else results.forEach(BookManagerApp::printBook);
                }
                case "6" -> {
                    int id = readInt("Введите ID книги для редактирования: ");
                    var opt = manager.findBook(id);
                    if (opt.isEmpty()) { System.out.println("Книга не найдена."); continue; }
                    Book old = opt.get();
                    System.out.println("Оставьте поле пустым, чтобы не менять.");
                    String newTitle = readString("Название (" + old.title() + "): ");
                    String newAuthor = readString("Автор (" + old.author() + "): ");
                    String newYearStr = readString("Год (" + old.year() + "): ");
                    String newGenre = readString("Жанр (" + old.genre() + "): ");
                    String newReadStr = readString("Статус (1-прочитана, 0-не прочитана) сейчас: " + (old.read() ? "1" : "0") + ": ");
                    String newRatingStr = readString("Оценка (" + old.rating() + "): ");
                    Map<String, Object> updates = new HashMap<>();
                    if (!newTitle.isBlank()) updates.put("title", newTitle);
                    if (!newAuthor.isBlank()) updates.put("author", newAuthor);
                    if (!newYearStr.isBlank()) {
                        try { updates.put("year", Integer.parseInt(newYearStr)); }
                        catch (NumberFormatException e) { System.out.println("Год должен быть числом, пропускаем."); }
                    }
                    if (!newGenre.isBlank()) updates.put("genre", newGenre);
                    if (!newReadStr.isBlank()) updates.put("read", newReadStr.equals("1"));
                    if (!newRatingStr.isBlank()) {
                        try { updates.put("rating", Integer.parseInt(newRatingStr)); }
                        catch (NumberFormatException e) { System.out.println("Оценка должна быть числом, пропускаем."); }
                    }
                    if (manager.editBook(id, updates)) System.out.println("Книга обновлена.");
                    else System.out.println("Ошибка обновления.");
                }
                case "7" -> {
                    int id = readInt("Введите ID книги для удаления: ");
                    if (manager.deleteBook(id)) System.out.println("Книга удалена.");
                    else System.out.println("Книга не найдена.");
                }
                case "8" -> {
                    var stats = manager.getStats();
                    System.out.println("\n=== СТАТИСТИКА ===");
                    System.out.println("Всего книг: " + stats.get("total"));
                    System.out.println("Прочитано: " + stats.get("read"));
                    System.out.println("Не прочитано: " + stats.get("unread"));
                    System.out.printf("Средняя оценка: %.2f%n", stats.get("avg_rating"));
                    System.out.println("По жанрам:");
                    @SuppressWarnings("unchecked")
                    Map<String, Integer> genres = (Map<String, Integer>) stats.get("genres");
                    genres.forEach((g, c) -> System.out.println("  " + g + ": " + c));
                }
                case "9" -> {
                    try {
                        manager.saveToFile("books_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "10" -> {
                    try {
                        manager.loadFromFile("books_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}
